### PR TITLE
Timer implementation

### PR DESCRIPTION
**Pull request by MykolaSolyanko**

### CODE REVIEW

**Review by Eabramov84 — APPROVED** [review]

File: include/aos/common/tools/timer.hpp [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+private:
+    static constexpr int cTimerSigevNotify = AOS_CONFIG_TIMER_SIGEV_NOTIFY;
+
+    static void TimerFunction(union sigval arg) { (static_cast<Timer*>(arg.sival_ptr)->mFunction)(); }
+
+    timer_t                                 mTimerID {};
+    unsigned int                            mIntervalMs {};
+    StaticFunction<cDefaultFunctionMaxSize> mFunction;
+};
+} // namespace aos
```
Comment (al1img) [inline include/aos/common/tools/timer.hpp:134]: Add empty line befor namespace closing.

File: include/aos/common/tools/timer.hpp [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+        sev.sigev_notify = cTimerSigevNotify;
+        sev.sigev_value.sival_ptr = this;
+        sev.sigev_notify_function = TimerFunction;
+
+        its.it_value.tv_sec = intervalMs / 1000;
+        its.it_value.tv_nsec = (intervalMs % 1000) * 1000000;
+
+        auto ret = timer_create(CLOCK_MONOTONIC, &sev, &mTimerID);
+        if (ret != 0) {
+            return ret;
```
Comment (al1img) [inline include/aos/common/tools/timer.hpp:72]: According to documentaion `timer_create` returns -1 in error case. I guess, errno should be returned.

File: include/aos/common/tools/timer.hpp [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        }
+
+        ret = timer_settime(mTimerID, 0, &its, nullptr);
```
Comment (al1img) [inline include/aos/common/tools/timer.hpp:75]: ditto

File: include/aos/common/tools/timer.hpp [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+    /**
+     * Stops timer.
+     *
+     * @return Error code.
+     */
+    Error Stop()
+    {
+        if (mTimerID != 0) {
+            auto ret = timer_delete(mTimerID);
```
Comment (al1img) [inline include/aos/common/tools/timer.hpp:91]: ditto

File: tests/common/src/tools/timer_test.cpp [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+#include <gtest/gtest.h>
+
+#include "aos/common/tools/timer.hpp"
+
+using namespace aos;
+
+TEST(common, TimerCreateAndStop)
+{
+
+    auto       g_interrupted = 0;
```
Comment (al1img) [inline tests/common/src/tools/timer_test.cpp:17]: it is not gloabal variable. Just `interrupted`.

File: tests/common/src/tools/timer_test.cpp [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    sleep(1);
+
+    EXPECT_TRUE(timer.Stop().IsNone());
+
+    EXPECT_EQ(1, g_interrupted);
+}
+
+TEST(common, TimerRaisedOnlyOnce)
+{
+    auto       g_interrupted = 0;
```
Comment (al1img) [inline tests/common/src/tools/timer_test.cpp:31]: ditto

File: tests/common/src/tools/timer_test.cpp [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    sleep(2);
+
+    EXPECT_TRUE(timer.Stop().IsNone());
+
+    EXPECT_EQ(1, g_interrupted);
+}
+
+TEST(common, TimerCreateResetStop)
+{
+    auto       g_interrupted = 0;
```
Comment (al1img) [inline tests/common/src/tools/timer_test.cpp:45]: ditto

File: include/aos/common/tools/timer.hpp (hunk 1 above)
Comment (MykolaSolyanko) [reply include/aos/common/tools/timer.hpp:135]: Done

File: include/aos/common/tools/timer.hpp (hunk 2 above)
Comment (MykolaSolyanko) [reply include/aos/common/tools/timer.hpp]: Done

File: include/aos/common/tools/timer.hpp (hunk 3 above)
Comment (MykolaSolyanko) [reply include/aos/common/tools/timer.hpp:75]: Done

File: include/aos/common/tools/timer.hpp (hunk 4 above)
Comment (MykolaSolyanko) [reply include/aos/common/tools/timer.hpp:91]: Done

File: tests/common/src/tools/timer_test.cpp (hunk 5 above)
Comment (MykolaSolyanko) [reply tests/common/src/tools/timer_test.cpp]: Done

File: tests/common/src/tools/timer_test.cpp (hunk 6 above)
Comment (MykolaSolyanko) [reply tests/common/src/tools/timer_test.cpp]: Done

File: tests/common/src/tools/timer_test.cpp (hunk 7 above)
Comment (MykolaSolyanko) [reply tests/common/src/tools/timer_test.cpp]: Done

**Review by al1img — APPROVED** [review]

**Pull request merged**